File: cpp/problem107.cpp
#include "mystd.h"
#include "BinaryTree.h"
using namespace std;
using namespace MYSTD;
typedef MYDS::BinaryTreeNode<int> TreeNode;

/**
 * simply apply BFS and then reverse the result
 *
 * Thanks for the moving paradigm in C++, the swap() operation
 * in reverse() has constant performance. Otherwise in an language
 * that always performs deep-copy, you may need to consider the
 * complexity in reverse.
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int> > res;
        levelOrder(root, res);
        reverse( res.begin(), res.end() );
        return res;
    }
    void levelOrder(TreeNode *root, vector<vector<int> > &res){
        queue< TreeNode *> sq;cout << res.size() << endl;
        if( root ) sq.emplace(root);
        while( !sq.empty() ){
            res.emplace_back();
            int sz = sq.size();
            for (int i = 0; i < sz; ++i){
                auto &front = sq.front();
                res.back().push_back(front->val);
                if( front->left ) sq.emplace(front->left);
                if( front->right ) sq.emplace(front->right);
                sq.pop();
            }
        }
    }
};

int main(int argc, char const *argv[]){
    // test by the example in Leetcode problem
    vector<int> v = {3,9,20,-1,-1,15,7};
    MYDS::BinaryTree<int> tr(v, -1);
    Solution sol;
    auto res = sol.levelOrderBottom( tr.root() );
    for(auto & i: res){
        MYFILE::print_range(i) << endl;
    }

    return 0;
}

File: swift/problem107.swift
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init(_ val: Int) {
        self.val = val
        self.left = nil
        self.right = nil
    }
}

//Breadth-first traversal
class Solution {
    func levelOrderBottom(_ root: TreeNode?) -> [[Int]] {
        if (root == nil) {
            return [[Int]]()
        }
        typealias node_dp = (node: TreeNode?, dp: Int)
        var queue = [node_dp]()
        var ans = [[Int]]()
        
        queue.append((root,0))
        
        while !queue.isEmpty {
            let tmp = queue.removeFirst()
            if (tmp.dp >= ans.count) {
                ans.append([tmp.node!.val])
            } else {
                ans[tmp.dp].append(tmp.node!.val)
            }
            if (tmp.node!.left != nil) {
                queue.append((tmp.node!.left,tmp.dp+1))
            }
            if (tmp.node!.right != nil) {
                queue.append((tmp.node!.right,tmp.dp+1))
            }
        }
        
        ans.reverse()
        return ans
    }
}
